File: src/commands/git/sync.ts
import * as prompts from "@clack/prompts";
import { Command } from "commander";
import { getCurrentBranch, getDefaultBranch } from "../../utils/git";

export const sync = new Command("sync")
	.description("Sync default branch and return to current branch")
	.action(async () => {
		prompts.intro("git sync");

		const s = prompts.spinner();
		try {
			const defaultBranch = await getDefaultBranch();
			const currentBranch = await getCurrentBranch();

			if (currentBranch === defaultBranch) {
				s.start("Pulling latest changes...");
				await Bun.$`git pull`.quiet();
				s.stop(`${defaultBranch} is up to date`);
			} else {
				s.start(`Switching to ${defaultBranch} and pulling...`);
				await Bun.$`git checkout ${defaultBranch}`.quiet();
				await Bun.$`git pull`.quiet();
				s.stop(`${defaultBranch} is up to date`);

				s.start(`Switching back to ${currentBranch}...`);
				await Bun.$`git checkout ${currentBranch}`.quiet();
				s.stop(`Back on ${currentBranch}`);
			}

			prompts.outro("Synced!");
		} catch (e: any) {
			s.stop("Failed");
			prompts.log.error(e.message);
			process.exit(1);
		}
	});

File: src/commands/git/sync.rs
use crate::utils::git::{Shell, get_current_branch, get_default_branch};
use anyhow::Result;

fn run_cmd(args: &[&str]) -> Result<String> {
    let output = std::process::Command::new("git").args(args).output()?;
    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        return Err(anyhow::anyhow!("git failed: {stderr}"));
    }
    Ok(String::from_utf8_lossy(&output.stdout).to_string())
}

pub fn execute_sync(shell: &dyn Shell) -> Result<()> {
    let default = get_default_branch(shell)?;
    let current = get_current_branch(shell)?;

    if current == default {
        println!("Pulling latest changes...");
        run_cmd(&["pull"])?;
        println!("{default} is up to date");
    } else {
        println!("Switching to {default} and pulling...");
        run_cmd(&["checkout", &default])?;
        run_cmd(&["pull"])?;
        println!("{default} is up to date");

        println!("Switching back to {current}...");
        run_cmd(&["checkout", &current])?;
        println!("Back on {current}");
    }

    println!("Synced!");
    Ok(())
}
